GRU.lua

```diff
@@ -0,0 +1,252 @@
+end
+
+
+function layer:reset(std)
+  if not std then
+    std = 1.0 / math.sqrt(self.hidden_dim + self.input_dim)
+  end
+  --self.bias:zero()
+  self.bias:normal(0,std) --self.bias[{{self.hidden_dim + 1, 2 * self.hidden_dim}}]:fill(1)
+  self.weight:normal(0, std)
+  return self
+end
+
+
+function layer:resetStates()
+  self.h0 = self.h0.new()
+end
+
+
+local function check_dims(x, dims)
+  assert(x:dim() == #dims)
+  for i, d in ipairs(dims) do
+    assert(x:size(i) == d)
+  end
+end
+
+
+function layer:_unpack_input(input)
+  local h0, x = nil, nil
+  
+  if torch.type(input) == 'table' and #input == 2 then
+    h0, x = unpack(input)
+  elseif torch.isTensor(input) then
+    x = input
+  else
+    assert(false, 'invalid input')
+  end
+  return h0, x
+end
+
+
+function layer:_get_sizes(input, gradOutput)
+  local h0, x = self:_unpack_input(input)
+  local N, T = x:size(1), x:size(2)
+  local H, D = self.hidden_dim, self.input_dim
+  check_dims(x, {N, T, D})
+  if h0 then
+    check_dims(h0, {N, H})
+  end
+  
+  if gradOutput then
+    check_dims(gradOutput, {N, T, H})
+  end
+  return N, T, D, H
+end
+
+
+--[[
+Input:
+- h0: Initial hidden state, (N, H)
+- x: Input sequence, (N, T, D)
+
+Output:
+- h: Sequence of hidden states, (N, T, H)
+--]]
+
+
+function layer:updateOutput(input)
+  local h0, x = self:_unpack_input(input)
+  local N, T, D, H = self:_get_sizes(input)
+
+  self._return_grad_h0 = (h0 ~= nil)
+  
+  if not h0 then
+    h0 = self.h0
+    if h0:nElement() == 0 or not self.remember_states then
+      h0:resize(N, H):zero()
+    elseif self.remember_states then
+      local prev_N, prev_T = self.output:size(1), self.output:size(2)
+      assert(prev_N == N, 'batch sizes must be the same to remember states')
+      h0:copy(self.output[{{}, prev_T}])
+    end
+  end
+
+  local bias_expand = self.bias:view(1, 3 * H):expand(N, 3 * H)
+  local Wx = self.weight[{{1, D}}]
+  local Wh = self.weight[{{D + 1, D + H}}]
+
+  local h = self.output
+  h:resize(N, T, H):zero()
+  local prev_h = h0
+  self.gates:resize(N, T, 3 * H):zero()
+  for t = 1, T do
+    local cur_x = x[{{}, t}]
+    local next_h = h[{{}, t}]
+    local cur_gates = self.gates[{{}, t}]
+    cur_gates:addmm(bias_expand, cur_x, Wx)
+    cur_gates[{{}, {1, 2 * H}}]:addmm(prev_h, Wh[{{}, {1, 2 * H}}])
+    cur_gates[{{}, {1, 2 * H}}]:sigmoid()
+    
+    local u = cur_gates[{{}, {1, H}}] --update gate : u = sig(Wx * x + Wh * prev_h + b)
+    local r = cur_gates[{{}, {H + 1, 2 * H}}] --reset gate : r = sig(Wx * x + Wh * prev_h + b)
+    next_h:cmul(r, prev_h) --temporary buffer : r . prev_h
+    cur_gates[{{}, {2 * H + 1, 3 * H}}]:addmm(next_h, Wh[{{}, {2 * H + 1, 3 * H}}]) -- hc += Wh * r . prev_h
+    local hc = cur_gates[{{}, {2 * H + 1, 3 * H}}]:tanh() --hidden candidate : hc = tanh(Wx * x + Wh * r . prev_h + b)
+    next_h:addcmul(prev_h,-1, u, prev_h)
+    next_h:addcmul(u,hc)  --next_h = (1-u) . prev_h + u . hc   
```

Comment:
It is true.
As always, there are many small variations for the same algorithm.
For the definition of GRU, I used the code in Karpathy's char-rnn and I didn't chek the original article.